### ISSUE TITLE
Do not make views

### ISSUE THREAD

**Issue by gasagna** [The-ReSolver/Fields.jl#13]
https://github.com/The-ReSolver/Fields.jl/blob/70e78c98be6fb942e7cefb6b75d85d47981708e9/src/derivatives.jl#L15

You should use the `mul!` method of `DiffMatrices` for 3D arrays.

**Comment by gasagna** [comment]
Just checked and using the `mul!` method for 3D arrays gives a speed-up of 10 on `update_r!`. :)

**Comment by gasagna** [comment]
If you want, you can have a method that would work for (dense) chebychev matrices, which does not have the nice `mul!` method for 3D arrays.

**Comment by gasagna** [comment]
Including threading and fixing the `mul!` method gives these times on my 4 CPU mac

```
(base) davide@vega - NSOperators.jl$: julia -t 8 perf/perf_update_cache.jl 
  8.812 ms (326 allocations: 26.28 KiB)
  2.610 ms (2116 allocations: 2.35 MiB)
  14.855 ms (392 allocations: 33.58 KiB)
```
while before it was
```
(base) davide@vega - NSOperators.jl$: julia -t 8 perf/perf_update_cache.jl 
  308.362 ms (24 allocations: 1.88 KiB)
  50.084 ms (2116 allocations: 2.35 MiB)
  319.173 ms (0 allocations: 0 bytes)
```

so a speed up of 35x, 19x and 22x! Not bad!

**Comment by tb6g16** [comment]
Ok, I'm thinking of a method that would get called inside ddy! like this:
```
function _diffy!(dudy::SpectralField{Ny, Nz, Nt, G, T}, Dy::DiffMatrix{T}, u::SpectralField{Ny, Nz, Nt, G, T}) where {Ny, Nz, Nt, G, T}
    LinearAlgebra.mul!(dudy, Dy, u)
end
```
However, this required DiffMatrix to be installed as a package. Is it possible to make DiffMatrix a subtype of something other than AbstractArray as this would allow something like
```
function _diffy!(dudy::SpectralField{Ny, Nz, Nt, G, T}, Dy<:D u::SpectralField{Ny, Nz, Nt, G, T}) where {Ny, Nz, Nt, G, T}
    LinearAlgebra.mul!(dudy, Dy, u)
end
```
where the type `D` is something like D<:AbstractSparseMatix or AbstractDiffMatrix? If it's too much work then it's not a problem, it's just a slightly cleaner implementation.

An another point, the method signature used for LinearAlgebra.mul! in FDGrids is
```
function LinearAlgebra.mul!(y::DenseArray{S, 3},
                            A::DiffMatrix{T, $WIDTH},
                            x::DenseArray{S, 3}) where {T, S}
```
Is it possible make the method work for abstract arrays in general? Or is there a good reason DenseArrays are specifically used?

**Comment by tb6g16** [comment]
An amendment to the last comment I made, if I use the parent of the arrays passed to mul! then there won't be a method error.

**Comment by gasagna** [comment]
Best thing would be to relax the types in `mul!` to be `AbstractArray`. I do not remember why I use `DenseArray`.

But you need to check that passing a `SpectralField` is not slower than passing its `parent`.

**Comment by tb6g16** [comment]
So to round out this issue. I've altered the method signature of mul! for `DiffMatrix` to work on `AbstractArray` instead of Dense arrays, and I updated ChebUtils so that it also has a similar interface for mul!. This means ddy! calls the correct method for `mul!` depending on if a `DiffMatrix` or `ChebDiff` is passed.

The `parent` of `SpectralField` does not need to be called anymore (because of the updated method signature), so the performance considerations are not a concern.
